File: hadoop-utils/src/main/scala/io/github/ppdzm/utils/hadoop/kafka/config/PropertiesBuilder.scala
package io.github.ppdzm.utils.hadoop.kafka.config

import _root_.io.github.ppdzm.utils.universal.base.Logging
import io.github.ppdzm.utils.universal.implicits.BasicConversions._
import org.apache.kafka.common.config.AbstractConfig

import java.util
import java.util.Properties
import scala.collection.JavaConversions._
import scala.collection.JavaConverters._
import scala.reflect._

/**
 * Created by Stuart Alex on 2021/1/28.
 */
private[config] abstract class PropertiesBuilder[T <: AbstractConfig : ClassTag] {
    private lazy val instance = null.asInstanceOf[T]
    // AbstractConfig子类的所有已CONFIG结尾的静态字段
    private lazy val fields = classTag[T].runtimeClass.getFields.filter(_.getName.endsWith("CONFIG"))
    // Builder类的方法
    private lazy val builderMethods = getClass.getMethods
    protected val properties = new Properties()
    private val logging = new Logging(getClass)

    def build(): Properties = properties

    def put(key: String, value: AnyRef): this.type = {
        properties.put(key, value)
        this
    }

    def invoke(parameters: Map[String, AnyRef]): this.type = {
        if (parameters != null)
            invoke(parameters.asJava)
        this
    }

    def invoke(parameters: util.Map[String, AnyRef]): this.type = {
        if (parameters == null)
            return this
        parameters.foreach {
            case (key, value) =>
                // 通过配置名称找到字段
                val fieldOption = fields.find(_.get(instance) == key)
                if (fieldOption.isDefined) {
                    this.logging.logInfo(s"try find method matches config field ${fieldOption.get.getName.red} ${"with value".green} ${key.red}")
                    // 通过字段名称找方法
                    invoke(fieldOption.get.getName.trimEnd("_CONFIG"), value)
                }
                else {
                    this.logging.logWarning(s"config filed with value ${key.red} ${"not found".yellow}")
                }
        }
        this
    }

    def invoke(methodName: String, parameter: AnyRef): this.type = {
        val methodOption = builderMethods.find(_.getName == methodName)
        if (methodOption.isDefined) {
            this.logging.logInfo(s"method named ${methodName.red} ${"found".green}")
            methodOption.get.invoke(this, parameter)
        } else {
            this.logging.logWarning(s"method named ${methodName.red} ${"not found".green}")
        }
        this
    }

}

File: hadoop-utils/src/main/java/io/github/ppdzm/utils/hadoop/kafka/config/PropertiesBuilder.java
package io.github.ppdzm.utils.hadoop.kafka.config;

import io.github.ppdzm.utils.universal.base.Logging;
import org.apache.kafka.common.config.AbstractConfig;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.util.Map;
import java.util.Properties;

public abstract class PropertiesBuilder<T extends AbstractConfig> {
    protected final Properties properties = new Properties();
    private final Method[] methods = getClass().getMethods();
    private final Logging logging = new Logging(getClass());
    private final Field[] fields = (((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0]).getClass().getFields();
    private final T instance = (T) null;

    public Properties build() {
        return properties;
    }

    public PropertiesBuilder<T> put(String key, Object value) {
        properties.put(key, value);
        return this;
    }

    public PropertiesBuilder<T> invoke(String method, Object value) throws InvocationTargetException, IllegalAccessException {
        boolean found = false;
        for (Method m : this.methods) {
            if (m.getName().equals(method)) {
                this.logging.logInfo("method named " + method + " found");
                m.setAccessible(true);
                m.invoke(this, value);
                found = true;
                break;
            }
        }
        if (!found) {
            this.logging.logWarning("method named " + method + " not found");
        }
        return this;
    }

    public PropertiesBuilder<T> invoke(Map<String, Object> parameters) throws InvocationTargetException, IllegalAccessException {
        if (parameters != null) {
            for (String key : parameters.keySet()) {
                Object value = parameters.get(key);
                boolean found = false;
                for (Field field : this.fields) {
                    field.setAccessible(true);
                    if (field.get(instance) == key) {
                        this.logging.logInfo("try find method matches config field " + field.getName() + " with value " + key);
                        invoke(field.getName().replace("_CONFIG", ""), value);
                        found = true;
                    }
                }
                if (!found) {
                    this.logging.logWarning("config filed with value " + key + " not found");
                }
            }
        }
        return this;
    }

}
